lib/pbench/cli/server/shell.py

```diff
@@ -37,6 +43,63 @@ def find_the_unicorn(logger: Logger):
             )
 
 
+def keycloak_connection(
+    config: PbenchServerConfig, logger: Logger
+) -> Optional[NoReturn]:
+    """
+     Checks if the Keycloak server is up and accepting the connections.
+     The connection check does the GET request on the oidc server /health
+     endpoint and the sample response returned by the /health endpoint looks
+     like the following:
+        {
+            "status": "UP", # if the server is up
+            "checks": []
+        }
+     Note: The Keycloak server needs to be started with health-enabled on.
+     Args:
+        config: PbenchServerConfig
+        logger: logger
+    Returns:
+        0 if successful
+    """
+    try:
+        oidc_server = config.get(
+            "authentication",
+            "internal_server_url",
+            fallback=config.get("authentication", "server_url"),
+        )
+    except (NoSectionError):
+        logger.exception("Bad config file:  missing 'authentication' section")
+        sys.exit(1)
+    except (NoOptionError):
+        logger.error("Bad config file:  no 'server_url' in  'authentication' section")
+        sys.exit(1)
+
+    session = requests.Session()
+    # The connection check will retry multiple times unless successful, viz.,
+    # [0.0s, 4.0s, 8.0s, 16.0s, ..., 120.0s]. urllib3 will sleep for:
+    # {backoff factor} * (2 ^ ({number of total retries} - 1)) seconds between
+    # the retries. However, the sleep will never be longer than backoff_max
+    # which defaults to 120.0s More detailed documentation on Retry and
+    # backoff_factor can be found at:
+    # https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry
+    try:
```

Comment:
Can we consider a slightly modified approach?  Below is a rework of this code to:

 1. Only wrap the GET operation with a try block so that we don't consider other errors as an OIDC connection problem
 2. Consistently use OIDC in the emitting log messages
 3. Move `sys.exit()` calls to `main()`
 4. Move return statements outside of the `try` block

Is it possible to get an 200 status code on the GET but the `status` field is not `UP`?  If so, I think we'd want to loop until we see `UP`.

```
def foo():
    retry = Retry(total=8, backoff_factor=2)
    adapter = HTTPAdapter(max_retries=retry)
    session.mount("http://", adapter)
    try:
        response = session.get(f"{oidc_server}/health")
    except Exception:
        logger.exception("Error connecting to the OIDC server")
        return 1
    if response.status_code == HTTPStatus.OK and response.json()["status"] == "UP":
        logger.debug("OIDC server connection verified")
        ret_val = 0
    else:
        logger.error(
            "OIDC client not running, OIDC server response: {}", response.json()
        )
        ret_val = 1
    return ret_val

def main():
    .
    .
    .
    ret_val = keycloack_connection(...)
    if ret_val != 0:
        sys.exit(ret_val)
```

Reply:
Yes, that's a good point.